### ISSUE TITLE
Document what to do with copy-pasted code

### ISSUE THREAD

**Issue by m-ou-se** [rust-lang/std-dev-guide#4]
It happens sometimes that code gets submitted for inclusion in the standard library which was taken (almost) literally from e.g. `itertools` or another library. We should document what to do with these w.r.t. copyright and crediting the original author(s).

**Comment by KodrAus** [comment]
Some of our modules have headings like this in them:

```
// Original implementation taken from rust-memchr.
// Copyright 2015 Andrew Gallant, bluss and Nicolas Koch
```

**Comment by m-ou-se** [comment]
Yeah, it'd be good to clarify if that happened with permission, and/or what the license of the original code was. Also, it would be nice to set the git author (or the Co-authored-by field) to the original author(s) when adding copied code, so they get some credit for it too.

For example, the files stat start with that snippet you just mentioned, were originally added in a commit with someone else as the git committer and author.

**Comment by KodrAus** [comment]
We could, say, expect at least the notice in the source itself and recommend the co-authored-by field along with a link back to the original source in the commit?

For contributors who aren't really familiar with git we might want a bit of a "how to use git in the Rust codebase" to help with rebases, squashing, co-authored-by, etc.
